### ISSUE TITLE
Overlay network not found on worker node

### ISSUE THREAD

**Issue by rvernica** [docker/compose#4399]
Docker Engine is running in swarm mode. I manually created an overlay attachable network `foo` on the *manager* node:
```
> docker network create --driver overlay --attachable foo
> docker network ls
NETWORK ID          NAME                DRIVER              SCOPE
b9653fe43369        bridge              bridge              local
f111200c51a0        docker_gwbridge     bridge              local
a0bdd5df7a75        host                host                local
6u4dleo7zzmn        ingress             overlay             swarm
e9de752139c3        none                null                local
i5rqhfn493w7        foo                 overlay             swarm
```
In the Docker Compose file I use:
```
networks:
  default:
    external:
      name: foo
```
Docker Compose services on the *manager* node start and attach to the network successfully.

Docker Compose services on the *worker* node **cannot** find the network:
```
> docker-compose up -d
WARNING: The Docker Engine you're using is running in swarm mode.

Compose does not use swarm mode to deploy services to multiple nodes in a swarm. All containers will be scheduled on the current node.

To deploy your application across the swarm, use `docker stack deploy`.

ERROR: Network foo declared as external, but could not be found. Please create the network manually using `docker network create foo` and try again.
```
Using `docker run` on the *worker* node, I can successfully attach a container to the overlay network:
```
> docker run --net foo bar
```
---
docker-engine: 1.13.0
docker-compose: 1.10.0

**Comment by lenovouser** [comment]
Same issue here

**Comment by PedroMD** [comment]
docker-engine: version 1.13.1, build 092cba3
docker-compose: version 1.11.1, build 7c5d5e4

using a compose file with a 'v3' schema

**Comment by xdexter** [comment]
+1

**Comment by myMartek** [comment]
The problem is that on the worker node there is no network called foo when calling `docker network ls`

The network becomes visible when you start some service like:
docker run -itd --name containerX --net foo busybox

as soon at least one container is running in that network it is visible in `docker network ls` and you can run docker-compose up with that external network

The "fix" for this issue could be:

1: Add a docker command to list all networks (eg. `docker network ls -a` in the swarm to be able to detect if the external network exists on a manager node
2. Ignore non existing networks in docker-compose when docker is running as a worker in a swarm and try to launch a service. Only if that fails forward the proper error

I would recommend solution 1 since this would be the most stable solution

**Comment by krabradosty** [comment]
Still not solved

**Comment by sakeven** [comment]
+1

**Comment by soeren-helbig** [comment]
+1

**Comment by westsouthnight** [comment]
=(

**Comment by agree2012** [comment]
++ (

**Comment by ktiwary2** [comment]
+1

**Comment by zhim** [comment]
/usr/local/lib/python3.5/dist-packages/compose/network.py
    def ensure(self):
        if self.external:
            if self.driver == 'overlay':
                # Swarm nodes do not register overlay networks that were
                # created on a different node unless they're in use.
                # See docker/compose#4399
                return
this is also not solve the question。

/usr/local/lib/python3.5/dist-packages/compose/config/config.py
 if external:
                validate_external(entity_type, name, config, config_file.version)
                if isinstance(external, dict):
                    config['name'] = external.get('name')
                elif not config.get('name'):

def validate_external(entity_type, name, config, version):
    if (version < V2_1 or (version >= V3_0 and version < V3_4)) and len(config.keys()) > 1:
        raise ConfigurationError(
            "{} {} declared as external but specifies additional attributes "
            "({}).".format(
                entity_type, name, ', '.join(k for k in config if k != 'external')))

                    config['name'] = name
external ‘s keys just one.

i consider

/usr/local/lib/python3.5/dist-packages/compose/network.py
    def ensure(self):
        if self.external:
            return

**Comment by shin-** [comment]
I answered that 4 days ago here: https://github.com/docker/compose/issues/4601#issuecomment-410881843

**Comment by dsyorkd** [comment]
Is there a better place to diagnose this error then? `docker4 dockerd[7293]: time="2020-05-14T23:38:16.283488926Z" level=error msg="fatal task error" error="starting container failed: failed to get network during CreateEndpoint: network tet6lnn6ntscd2ismuidagxix not found" module=node/agent/taskmanager node.id=ka0dhbpjg2rylj6t9n74z6n7b service.id=4saanfb0myoubp2i5phcbvib2 task.id=4v21u79tvfns6buoet74fkae9` 

For the better part of the day I've been looking through logs and inspecting containers, nodes, etc to figure out why these containers would be failing on this worker node, the node is communicating with the cluster from most accounts I can tell, tasks are being scheduled and containers not relying on the overlay are starting up.

**Comment by dominiwe** [comment]
Still not fixed in v[2.21.0](https://docs.docker.com/compose/release-notes/#2210)...
